### PR TITLE
Add sidecar forensics utilities

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Adds a `RecordParsers.parseV6SidecarRecordsByConsTimeIn()` method to parse the side cars in a directory for correlation with the main record stream.

### CODE REVIEW

**Review by david-bakin-sl — COMMENTED** [review]
First time I've seen code like this in this codebase (just starting to look at it, of course) and I do _like_ the functional approach by filtering/transforming datastreams.  The Java syntax is somewhat unwieldy (vs, say, Haskell or OCaml or LINQ) but once you get used to it - and data pipelines at all! - its very readable.

**Comment by tinker-michaelj** [comment]
> First time I've seen code like this in this codebase (just starting to look at it, of course) and I do _like_ the functional approach by filtering/transforming datastreams. The Java syntax is somewhat unwieldy (vs, say, Haskell or OCaml or LINQ) but once you get used to it - and data pipelines at all! - its very readable.

Yeah it's less typing in a lot of case, though the Streams overhead ~always prohibits use in production code paths.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/forensics/RecordParsers.java [hunk 1]
```diff
@@ -74,12 +82,73 @@ public static List<RecordStreamEntry> parseV6RecordStreamEntriesIn(final String
...
+        return filteredFilesFrom(
+                streamDir,
+                s -> s.endsWith(V6_FILE_EXT) && !s.contains("Z_"),
+                RecordParsers::parseRecordFileConsensusTime);
+    }
+
+    private static List<String> orderedSidecarFilesFrom(final String streamDir) throws IOException {
+        return filteredFilesFrom(
+                streamDir,
+                s -> s.endsWith(V6_FILE_EXT) && s.contains("Z_"),
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/forensics/RecordParsers.java]: Added constant!

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java [hunk 2]
```diff
@@ -248,7 +247,77 @@ public List<HapiApiSpec> getSpecsInSuite() {
...
+                        sourcing(
+                                () ->
+                                        contractCall(
+                                                        APPROVE_BY_DELEGATE,
+                                                        "doIt",
+                                                        asHeadlongAddress(tiqMirrorAddr.get()),
+                                                        asHeadlongAddress(
+                                                                somebodyElseMirrorAddr.get()),
+                                                        new BigInteger(
+                                                                "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
```
[the remark this answers is not in the document]
Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java]: Added constant.

**Comment by david-bakin-sl** [comment]
> ...though the overhead of the Streams API ~always prohibits use in production code paths.

Don't get me started... (but for some other forum I've got a rant ready, the gist of it is that it doesn't have to have been this way ...  😒 )

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/forensics/RecordParsers.java [hunk 3]
```diff
@@ -35,6 +43,11 @@
  * runtime.
  */
 public class RecordParsers {
+    // A token that we can use to distinguish sidecar files with names like,
+    //   2022-12-05T14_23_46.192841556Z_01.rcd.gz
+    // from record stream files whose names do *not* include a _XX index
+    // like _01, _02, etc.
+    private static final String SIDECAR_ONLY_TOKEN = "Z_";
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/forensics/RecordParsers.java:50]: How about something closer to 'sidecar' like `"SCAR_"` or `"SC_"` (Not gonna lie, kinda like SCAR)

**Comment by mhess-swl** [comment]
I don't feel like I know enough about sidecars and contracts to approve this, but it does LGTM 😄

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/forensics/RecordParsers.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/utils/forensics/RecordParsers.java:50]: Ah I get your point...something like,
```
2022-12-05T14_23_46.192841556Z_SCAR_01.rcd.gz
```
yes that would have been nice...I think we have a contract with mirror nodes with this format now tho. 🤷

**Pull request merged**